cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -1209,6 +1660,75 @@ void RegisterScalarTemporal(FunctionRegistry* registry) {
           OutputType::Resolver(ResolveAssumeTimezoneOutput), &assume_timezone_doc,
           nullptr, AssumeTimezoneState::Init);
   DCHECK_OK(registry->AddFunction(std::move(assume_timezone)));
+
+  auto years_between = MakeTemporalBinary<YearsBetween, TemporalBinary, Int64Type>(
+      "years_between", {WithDates, WithTimestamps}, int64(), &years_between_doc);
+  DCHECK_OK(registry->AddFunction(std::move(years_between)));
+
+  auto quarters_between = MakeTemporalBinary<QuartersBetween, TemporalBinary, Int64Type>(
+      "quarters_between", {WithDates, WithTimestamps}, int64(), &quarters_between_doc);
+  DCHECK_OK(registry->AddFunction(std::move(quarters_between)));
+
+  auto month_interval_between =
+      MakeTemporalBinary<MonthsBetween, TemporalBinary, MonthIntervalType>(
+          "month_interval_between", {WithDates, WithTimestamps}, month_interval(),
+          &months_between_doc);
+  DCHECK_OK(registry->AddFunction(std::move(month_interval_between)));
+
+  auto month_day_nano_between =
+      MakeTemporalBinary<MonthDayNanoBetween, TemporalBinary, MonthDayNanoIntervalType>(
+          "month_day_nano_interval_between", {WithDates, WithTimes, WithTimestamps},
+          month_day_nano_interval(), &month_day_nano_interval_between_doc);
+  DCHECK_OK(registry->AddFunction(std::move(month_day_nano_between)));
+
+  auto weeks_between =
+      MakeTemporalBinary<WeeksBetween, TemporalDayOfWeekBinary, Int64Type>(
+          "weeks_between", {WithDates, WithTimestamps}, int64(), &weeks_between_doc,
+          &default_day_of_week_options, DayOfWeekState::Init);
+  DCHECK_OK(registry->AddFunction(std::move(weeks_between)));
+
+  auto day_time_between =
```

Comment:
Nit: `day_time_interval_between` would be more consistent. Same for `month_day_nano_interval_between`. I don't feel strongly about it, just noticing :).